util/types/binstring.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+// BinString is the internal type for storing bit / hex literal type.
+type BinString struct {
+	// Value holds the raw value for literal in BigEndian.
+	Value []byte
+}
+
+// BitLiteral is the bit literal type.
+type BitLiteral struct {
+	BinString
+}
+
+// HexLiteral is the hex literal type.
+type HexLiteral struct {
+	BinString
+}
+
+// ZeroBinString is a BinString literal with zero value.
+var ZeroBinString = BinString{[]byte{}}
+
+func trimLeadingZeroBytes(bytes []byte) []byte {
+	if len(bytes) == 0 {
+		return bytes
+	}
+	var pos int
+	posMax := len(bytes) - 1
+	for pos = 0; pos < posMax; pos++ {
+		if bytes[pos] != 0 {
+			break
+		}
+	}
+	return bytes[pos:]
+}
+
+// NewBinStringFromBytes creates a new BinString instance by the given bytes.
+// The leading zero in bytes are required to be trimmed before feeding to this function.
+func NewBinStringFromBytes(bytes []byte) BinString {
+	return BinString{bytes}
+}
+
+// NewBinStringFromUint creates a new BinString instance by the given uint value in BitEndian.
+// byteSize will be used as the length of the new BinString, with leading bytes filled to zero.
+// If byteSize is -1, the leading zeros in new BinString will be trimmed.
+func NewBinStringFromUint(value uint64, byteSize int) BinString {
+	if byteSize != -1 && (byteSize < 1 || byteSize > 8) {
+		panic("Invalid byteSize")
+	}
+	bytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(bytes, value)
+	if byteSize == -1 {
+		bytes = trimLeadingZeroBytes(bytes)
+	} else {
+		bytes = bytes[8-byteSize:]
+	}
+	return NewBinStringFromBytes(bytes)
+}
+
+// String implements fmt.Stringer interface.
+func (b BinString) String() string {
+	if len(b.Value) == 0 {
+		return ""
+	}
+	return "0x" + hex.EncodeToString(b.Value)
+}
+
+// ToString returns the string representation for the literal.
+func (b BinString) ToString() string {
```

Comment:
when should we use `String()` and `ToString()` ?

Reply:
String() is not supposed to be called directly. It is for fmt.Printf so that we can get a more friendly output.